gem/optimise.py

```diff
@@ -413,3 +474,758 @@ def aggressive_unroll(expression):
+                    d[key] = [factor]
+            rep.append(d)
+        self.rep = rep
+
+    def find_optimal_arg_factors(self):
+        # TODO: add guard so that self.rep is up to date here
+        rep = self.rep
+        arg_ind_flat = tuple([i for id in self.arg_ind for i in id])
+        arg_ind_set = set(self.arg_ind_flat)
+
+        gem_int = OrderedDict()  # Gem node -> int
+        int_gem = OrderedDict()  # int -> Gem node
+        extent = dict()  # stores product of extents of free indices
+        counter = 0
+        # TODO: perhaps should keep a list of all nodes in the object
+        # assign number to all factors
+        for summand in rep:
+            for key, factors in summand.items():
+                # TODO: this pattern appears multiple times, need to rewrite
+                if key == 'const' or key == 'quad':
+                    continue
+                for factor in factors:
+                    if factor not in gem_int:
+                        gem_int[factor] = counter
+                        int_gem[counter] = factor
+                        extent[counter] = numpy.product([i.extent for i in set(factor.free_indices) & arg_ind_set])
+                        counter += 1
+        if counter == 0:
+            return tuple()
+        # add connections (list of tuples)
+        connections = []
+        for summand in rep:
+            connection = []
+            for key, factors in summand:
+                if key == 'const' or key == 'quad':
+                    continue
+                connection.extend([gem_int[factor] for factor in factors])
+            connections.append(tuple(connection))
+
+        # set up the ILP
+        import pulp as ilp
+        ilp_prob = ilp.LpProblem('gem factorise', ilp.LpMinimize)
+        ilp_var = ilp.LpVariable.dicts('node', int_gem.keys(), 0, 1, ilp.LpBinary)
+
+        # Objective function
+        # Minimise number of factors to pull. If same number, favour factor with larger extent
+        big = 10000000  # some arbitrary big number
+        ilp_prob += ilp.lpSum(ilp_var[i] * (big - extent[i]) for i in int_gem)
+
+        # constraints
+        for connection in connections:
+            ilp_prob += ilp.lpSum(ilp_var[i] for i in connection) >= 1
+
+        ilp_prob.solve()
+        if prob.status != 1:
+            raise AssertionError("Something bad happened during ILP")
+
+        optimal_factors = [factor for factor, number in gem_int.items() if ilp_var[number].value() == 1]
+        other_factors = [factor for factor, number in gem_int.items() if ilp_var[number].value() == 0]
+        # TODO: investigate effects of sorting these two lists of factors
+        optimal_factors = sorted(optimal_factors, key = lambda f: extent[gem_int[f]])
+        other_factors = sorted(other_factors, key=lambda f: extent[gem_int[f]])
+        # Sequence dictating order of factorisation
+        factors_seq = optimal_factors + other_factors
+        return factors_seq
```

Comment:
It seems that you never use `'const'` or `'quad'` in this function, am I correct?